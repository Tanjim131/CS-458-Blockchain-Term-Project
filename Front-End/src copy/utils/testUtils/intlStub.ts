import { createIntl } from "react-intl";
const en_us = require("../../i18n/translations/en.json");

export const intlStub = createIntl({ locale: "en-us", messages: en_us });
export const intlMessages = en_us;
