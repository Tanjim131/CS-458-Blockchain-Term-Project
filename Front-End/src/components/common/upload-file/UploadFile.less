.document-details .ant-image {
  width: 100%;
  border: 1px solid #95989a;
  border-radius: 2px;
  background: #eeeeee;
  height: 96px;
  color: #9a9a9a;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 8px;
  padding: 3px;
  cursor: pointer;
  .ant-image-img {
    height: 90px;
    width: auto;
    max-width: 120px;
  }
}
.document-details {
  margin-top: 10px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  .remove-file {
    float: right;
    font-size: 24px;
    color: #f05a23;
  }
  .document-information {
    display: flex;
    align-items: center;
    .pdf-file-icon {
      font-size: 32px;
      color: #f05a23;
    }
    .document-name-size {
      display: flex;
      flex-direction: column;
      margin-left: 10px;
      .document-name {
        color: #282828;
        word-break: break-all;
      }
      .document-size {
        color: #788ca0;
      }
    }
  }
}
.upload-document-field {
  margin: 20px 0 5px 0;
}
.upload-file-error {
  color: #d24b19;
}
.s3-upload-file-error {
  border-color: #d24b19 !important;
}
.ant-upload-list-picture-card-container {
  display: none;
}
.s3-upload-file {
  border: 1px solid #d9d9d9;
  border-radius: 2px;
  background: #f9f9f9;
  height: 115px;
  color: #788ca0;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  &.disabled {
    pointer-events: none;
    opacity: 0.5;
  }
  &:hover {
    border: 1px solid #40a9ff;
  }
  input {
    display: none;
  }
  .upload-text {
    font-size: 14px;
    letter-spacing: 0.14px;
    color: #282828;
    font-weight: 600;
    margin-bottom: 8px;
    .upload-icon {
      margin-right: 8px;
    }
  }
}
