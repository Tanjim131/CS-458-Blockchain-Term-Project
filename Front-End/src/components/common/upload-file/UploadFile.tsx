import React, { useState, FC, useRef, useEffect } from "react";
import { Form, Typography, Row, Col, Image } from "antd";
import {
  InfoCircleOutlined,
  FilePdfOutlined,
  FileImageOutlined,
  DeleteOutlined,
  UploadOutlined,
} from "@ant-design/icons";
import { injectIntl, IntlShape } from "react-intl";
import generatePdfThumbnails from "pdf-thumbnails-generator";
import translate from "../../../../i18n/translate";
import byteToMB from "../../../../utils/file/byteToMB";
import { fileToBase64 } from "../../../../utils/file/fileToBase64";
import {
  getFileSizeInMB,
  getAllowedFileFormat,
} from "../../../../utils/file/downloadFromUrl";
import isMobile from "../../../../utils/type/isMobile";
import convertFileSize from "../../../../utils/file/fileSize";

import "./UploadFile.less";

const { Text } = Typography;
const S3_UPLOAD_FILE = "s3-upload-file ";
const S3_UPLOAD_FILE_ERROR = "s3-upload-file-error";

/**
 * Interface for File Details.
 */
interface IFileDetailsProps {
  fileSize?: number;
  fileName?: string;
  fileExtension?: string;
  thumbnailUrl?: string;
}
/**
 * Interface for upload file.
 */
interface IUploadFileProps {
  intl: IntlShape;
  fileRequired?: string;
  isClear: boolean;
  fileDetails?: IFileDetailsProps;
  disabled?: boolean;
  onFileUpload?: (file: any) => void;
  onRemoveFile?: () => void;
}

/**
 * This is used for upload file
 *
 * @usage: <UploadFile onFileUpload={} onRemoveFile={} fileUploadError={string} uploadLabel={string} uploadInfo={string} uploadInformation1={string} uploadInformation2={string} uploadInformation3={string} />
 */
const UploadFile: FC<IUploadFileProps> = (props: IUploadFileProps) => {
  const {
    intl,
    isClear,
    fileRequired = "",
    fileDetails = "",
    disabled = false,
    onFileUpload,
    onRemoveFile,
  } = props;
  const hiddenFileInput = useRef(null);
  const [uploadFileName, setFileName] = useState("");
  const [thumbnailImg, setThumbnailImg] = useState("");
  const [fileIcon, setFileIcon] = useState("");
  const [fileUploadError, setFileUploadError] = useState("");
  const [uploadFileSize, setfileSize] = useState(0);

  const APPLICATION_PDF = "application/pdf";

  useEffect(() => {
    if (isClear) {
      setFileName("");
      setfileSize(0);
      setThumbnailImg("");
      setFileUploadError("");
    }
  }, [isClear]);

  useEffect(() => {
    setFileUploadError(fileRequired);
  }, [fileRequired]);

  useEffect(() => {
    if (fileDetails) {
      if (fileDetails.fileSize) {
        setfileSize(fileDetails.fileSize);
      }
      if (fileDetails.fileName && fileDetails.fileExtension) {
        const setDocName: string = fileDetails.fileName.includes(
          fileDetails.fileExtension
        )
          ? fileDetails.fileName
          : `${fileDetails.fileName}.${fileDetails.fileExtension}`;
        setFileName(setDocName);
      }
      if (fileDetails.thumbnailUrl) {
        setThumbnailImg(fileDetails.thumbnailUrl);
      }
      if (fileDetails.fileExtension) {
        setFileIcon(fileDetails.fileExtension === "pdf" ? APPLICATION_PDF : "");
      }
    }
  }, [fileDetails]);
  /**
   * This is used for handle click for upload document
   */
  const handleFileUploadbtn = () => {
    hiddenFileInput.current.click();
  };
  /**
   * This is used for handle the upload document
   */
  const handleFileUpload = async (file: any) => {
    if (file && getAllowedFileFormat(file[0].type)) {
      setfileSize(file[0].size);
      const fileSize: any = byteToMB(file[0].size);
      if (fileSize > 2) {
        setFileUploadError(intl.formatMessage({ id: "upload_2mb_error" }));
      } else {
        setFileName(file[0].name);
        const imageStr: any = await fileToBase64(file[0]);
        if (file[0].type === APPLICATION_PDF) {
          const thumbnails = await generatePdfThumbnails(imageStr, 200);
          setThumbnailImg(thumbnails[0].thumbnail);
          file.thumbnails = thumbnails[0].thumbnail;
          setFileIcon(APPLICATION_PDF);
        } else {
          setThumbnailImg(imageStr);
          setFileIcon("");
          file.thumbnails = imageStr;
        }
        
        if(onFileUpload){
          onFileUpload(file);
        }

        setFileUploadError("");
      }
    } else {
      setFileUploadError(intl.formatMessage({ id: "accept_files_msg" }));
    }
  };
  /**
   * This is used for remove upload document
   */
  const removeUploadedFile = () => {
    setFileName("");
    setfileSize(0);
    setThumbnailImg("");
    setFileUploadError("");
    if(onRemoveFile){
      onRemoveFile();
    }
  };

  return (
    <>
      <Form.Item
        name="docFile"
        label={translate("upload_soft_copy_label")}
        className="add-document-form-field upload-document-field"
        tooltip={{
          title: translate("upload_soft_copy_info"),
          icon: <InfoCircleOutlined />,
          placement: "right",
          overlayClassName: "tooltip-add-document",
        }}
      >
        {thumbnailImg ? (
          <>
            <div className="document-details">
              <Image
                className="upload-pdf-img-preview"
                src={thumbnailImg}
              />
              <Row>
                <Col span={20} className="document-information">
                  {fileIcon ? (
                    <FilePdfOutlined className="pdf-file-icon" />
                  ) : (
                    <FileImageOutlined className="pdf-file-icon" />
                  )}
                  <Text className="document-name-size">
                    <Text className="document-name">{uploadFileName}</Text>
                    <Text className="document-size">
                      {convertFileSize(uploadFileSize)}
                    </Text>
                  </Text>
                </Col>
                <Col span={4}>
                  <DeleteOutlined
                    className="remove-file"
                    onClick={removeUploadedFile}
                  />
                </Col>
              </Row>
            </div>
          </>
        ) : (
          <>
            <div
              className={
                S3_UPLOAD_FILE +
                (fileUploadError ? S3_UPLOAD_FILE_ERROR : "") +
                (disabled ? " disabled" : "")
              }
              onClick={handleFileUploadbtn}
            >
              <div className="upload-text">
                <UploadOutlined className="upload-icon" />
                {isMobile()
                  ? translate("upload_from_mobile")
                  : translate("upload_from_desktop")}
              </div>
              <div>{translate("upload_information2")}</div>
              <div>{translate("upload_information3")}</div>
              <input
                ref={hiddenFileInput}
                type="file"
                name="files"
                accept="application/pdf, image/png, image/jpeg"
                onChange={(e) => handleFileUpload(e.target.files)}
                style={{ display: "none" }}
              />
            </div>
            {fileUploadError ? (
              <div className="upload-file-error">{fileUploadError}</div>
            ) : (
              ""
            )}
          </>
        )}
      </Form.Item>
    </>
  );
};

export default injectIntl(UploadFile);
