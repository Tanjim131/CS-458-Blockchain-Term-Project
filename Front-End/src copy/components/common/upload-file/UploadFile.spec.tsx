import React from 'react';
import { shallow } from "enzyme";
import UploadFile from "./UploadFile";
import { shallowToJson } from "enzyme-to-json";
import { intlStub } from '../../../../utils/testUtils/intlStub';

describe('UploadFile Component', () => {
    let wrapper: any;

    const fileDetail = {
        fileSize : 123,
        fileName : "name",
        fileExtension : ".pdf",
        thumbnailUrl : "https://jsonplaceholder.com"
    }
    beforeAll(() => {
        wrapper = shallow(
            <UploadFile.WrappedComponent
            intl = {intlStub}
            fileRequired = "required"
            isClear = {true}
            fileDetails = {fileDetail}
            disabled = {true}
            onFileUpload = {jest.fn()}
            onRemoveFile = {jest.fn()}
                      />
        )
    })

    it('Check if component exists or not and capture snapshot', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    })

})