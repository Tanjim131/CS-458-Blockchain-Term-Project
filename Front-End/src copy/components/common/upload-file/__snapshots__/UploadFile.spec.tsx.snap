// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadFile Component Check if component exists or not and capture snapshot 1`] = `
<Fragment>
  <FormItem
    className="add-document-form-field upload-document-field"
    label={
      <FormattedMessage
        id="upload_soft_copy_label"
        values={Object {}}
      />
    }
    name="docFile"
    tooltip={
      Object {
        "icon": <ForwardRef(InfoCircleOutlined) />,
        "overlayClassName": "tooltip-add-document",
        "placement": "right",
        "title": <FormattedMessage
          id="upload_soft_copy_info"
          values={Object {}}
        />,
      }
    }
  >
    <div
      className="s3-upload-file  disabled"
      onClick={[Function]}
    >
      <div
        className="upload-text"
      >
        <ForwardRef(UploadOutlined)
          className="upload-icon"
        />
        <FormattedMessage
          id="upload_from_desktop"
          values={Object {}}
        />
      </div>
      <div>
        <FormattedMessage
          id="upload_information2"
          values={Object {}}
        />
      </div>
      <div>
        <FormattedMessage
          id="upload_information3"
          values={Object {}}
        />
      </div>
      <input
        accept="application/pdf, image/png, image/jpeg"
        name="files"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
    </div>
  </FormItem>
</Fragment>
`;
